docs: document latest manual tracing (#20)
open-telemetry/opentelemetry-js-api: docs/tracing.md

# Tracing

This quick start is for end users of OpenTelemetry who wish to manually trace their applications. If you are a library author, please see the [Library Authors Guide](library-author.md). If you wish to automatically instrument your application, see the automatic instrumentation documentation for the SDK you wish to use.

_Trace API reference: <https://open-telemetry.github.io/opentelemetry-js/classes/traceapi.html>_

### Acquiring a Tracer

In OpenTelemetry, tracing operations are performed using methods on a _tracer_. You can get a tracer by calling [`getTracer`](https://open-telemetry.github.io/opentelemetry-js/classes/traceapi.html#gettracer) on the global tracer provider. `getTracer` takes the name and version of the application or library acquiring the tracer, and provides a tracer which can be used to trace operations.

```typescript

> import { trace } from '@opentelemetry/api';

Reviewer: I'm fine with using typescript in samples/docs but till now javascript is usually used everywhere.

Author: Only difference is `import` vs `require` for now. None of the code examples have types in them

const tracer = trace.getTracer("my-application", "0.1.0");
```

### Starting and Ending a Span

In OpenTelemetry, all _traces_ are composed of [`Spans`](https://open-telemetry.github.io/opentelemetry-js/interfaces/span.html). These spans are linked together by parent-child relationships to form a tree. The resultant tree is your trace, and the root of the tree is commonly called the _root span_.

You can create a span by calling [`Tracer#startSpan`](https://open-telemetry.github.io/opentelemetry-js/interfaces/tracer.html#startspan). The only required argument to `startSpan` is the _span name_, which should describe the operation being performed with low cardinality.

```typescript
const span = tracer.startSpan("my-span-name");

// do some work

// When a span is ended, it will be exported to a tracing backend
// via the currently registered SDK.
span.end();
```

Most of the time, spans will be used as part of a function which responds to some event like a web request. The following example shows what it might look like to manually trace a function which responds to a get request using an imaginary http server framework.

```typescript
async function onGet(request, response) {

>   const span = tracer.startSpan("GET");

Reviewer: I think it would be good to mention also `SpanKind` somewhere.

```

### Span Attributes

While name, start time, end time, and status are the minimum information required to trace an operation, most of the time they will not be enough information on their own to effectively observe an application. In the example above we traced a `GET` request, but the same span name would be used for every route which would make it impossible to know what route was called by looking at the trace. To solve this, OpenTelemetry uses _Span Attributes_. Span attributes are an object with string keys and string, number, or boolean values which describe the span. For example, we can use the span attributes to add route and http response code information to the example above.

```typescript
async function onGet(request, response) {
  const span = tracer.startSpan("GET", {
    // attributes can be added when the span is started
    attributes: {
      "http.route": request.route,
    }
  });

  // attributes may also be added after the span is started
  span.setAttribute("http.status", 200);
  
  span.end();

  // Attributes MAY NOT be added after the span ends
  span.setAttribute("my.attribute", false); // this does nothing
}
```

#### Semantic Conventions

> One problem with span names and attributes is recognizing, categorizing, and analyzing them in your tracing backend. Between different applications, libraries, and tracing backends there might be different names and expected values for various attributes. For example, your application may use `http.status` to describe the HTTP status code, but a library you use may use `http.status_code`. In order to solve this problem, OpenTelemetry uses a library of semantic conventions which describe the name and attributes which should be used for specific types of spans.

Reviewer: Maybe extend this a little bit. e.g. recommended usage where applicable but mention also that other keys are allowed.

_See the current trace semantic conventions in the OpenTelemetry Specification repository: <https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions>_

### Span Relationships

Now that we can start and end spans and effectively describe them, it is useful to be able to describe their relationships to each other. For instance, if one span describes an incoming request which makes a database call, it may be useful to trace the database call as a separate span which is a child of the original request span. In order to do this, when we create a span we can tell OpenTelemetry which span to use as its parent using a mechanism called _Context_.

Context is a very important part of the OpenTelemetry API which cannot be adequately explained in a single paragraph. To read more about context, see the [context documentation](context.md).

```typescript
import { context, setSpan } from '@opentelemetry/api';

async function onGet(request, response) {

>   const span = tracer.startSpan("GET /user/:id", {

Reviewer: should be most likely 
```typescript
const span = tracer.startSpan("GET /user/:id", {
  attributes: {
    "http.method": "GET",
    "http.url": request.url,
  }
});
```
I think we should use semantic conventions in our samples if applicable - and SpanKind.

    "http.method": "GET",
    "http.url": request.url,
  });

  const userId = request.params.id;

  // Create a new context from the current context which has the span "active"
  const ctx = setSpan(context.active(), span);
  
  // Call getUser with the newly created context

>   const user = await context.with(ctx, getUser, userId);

Reviewer: ```suggestion
  const user = await context.with(ctx, getUser, undefined, userId);
```
The first argument after the function is the receiver.

# Tracing

This quick start is for end users of OpenTelemetry who wish to manually trace their applications. If you are a library author, please see the [Library Authors Guide](library-author.md). If you wish to automatically instrument your application, see the automatic instrumentation documentation for the SDK you wish to use.

_Trace API reference: <https://open-telemetry.github.io/opentelemetry-js-api/classes/traceapi.html>_

## Acquiring a Tracer

In OpenTelemetry, tracing operations are performed using methods on a _tracer_. You can get a tracer by calling [`getTracer`](https://open-telemetry.github.io/opentelemetry-js-api/classes/traceapi.html#gettracer) on the global tracer provider. `getTracer` takes the name and version of the application or library acquiring the tracer, and provides a tracer which can be used to trace operations.

```typescript
import { trace } from '@opentelemetry/api';

const tracer = trace.getTracer("my-application", "0.1.0");
```

## Starting and Ending a Span

> In OpenTelemetry, all _traces_ are composed of [`Spans`](https://open-telemetry.github.io/opentelemetry-js/interfaces/span.html). These spans are linked together by parent-child relationships to form a tree. The resultant tree is your trace, and the root of the tree is commonly called the _root span_.

Reviewer: Maybe add `A span should describe a single operation like a database request, an incoming web request,...`.

      code: SpanStatusCode.ERROR,
      message: err.message,
    });
  } finally {
    // Every span must be ended or it will not be exported
    span.end();
  }
}

server.on("GET", "/user/:id", onGet);
```

## Describing a Span

Using span relationships, attributes, kind, and the related semantic conventions, we can more accurately describe the span in a way our tracing backend will more easily understand. The following example uses these mechanisms, which are described below.

```typescript
import { context, setSpan, SpanKind } from '@opentelemetry/api';

async function onGet(request, response) {
  // HTTP semantic conventions determine the span name and attributes for this span
  const span = tracer.startSpan(`GET /user/:id`, {
    // attributes can be added when the span is started
    attributes: {
      "http.method": "GET",
      "http.url": request.url
    },
    // This span represents a remote incoming synchronous request
    kind: SpanKind.SERVER
  });

  const userId = request.params.id;

  // Create a new context from the current context which has the span "active"
  const ctx = setSpan(context.active(), span);
  
  // Call getUser with the newly created context

>   const user = await context.with(ctx, getUser, undefined, userId);

Reviewer 2: i think we could inline the callback here, the fact that we are giving its param with the 4th arg is kinda complex to understand for beginner i think

Author: I think if a user has trouble with this then the tsdoc needs to be made more clear. This is the pattern I expect most users to use, not just advanced ones.

Reviewer 2: Do you mean adding an exemple inside the tsdoc directly ? I'm fine with that, i think this would be much clearer:
```ts
const userId = request.params.id;
// Create a new context from the current context which has the span "active"
const ctx = setSpan(context.active(), span);
// Call getUser with the newly created context
const user = await context.with(ctx, async () => {
  // when this span is created, it will automatically use the span from the context as its parent
  const span = tracer.startSpan("SELECT ShopDb.Users", {
    attributes: {
      // Attributes from the database trace semantic conventions
      // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md
      "db.system": "mysql",
      "db.connection_string": "Server=shopdb.example.com;Database=ShopDb;Uid=billing_user;TableCache=true;UseCompression=True;MinimumPoolSize=10;MaximumPoolSize=50;",
      "db.user": "app_user",
      "net.peer.name": "shopdb.example.com",
      "net.peer.ip": "192.0.2.12",
      "net.peer.port": 3306,
      "net.transport": "IP.TCP",
      "db.name": "ShopDb",
      "db.statement": `Select * from Users WHERE user_id = ${userId}`,
      "db.operation": "SELECT",
      "db.sql.table": "Users",
    },
    kind: SpanKind.CLIENT,
  });
  const user = await db.select("Users", { id: userId });
  span.end();
  return user;
})
```

WDYT ?

Author: What if I annotate the call to context.with like this? I can change to the inline style if you still think that's more clear, but I think this calling convention will be the more common and recommended approach most of the time.

```javascript
  // Call getUser with the newly created context
  // 
  // context.with calls a function with an associated "active" context. Within
  // the function, calling context.active() returns the currently active context.
  // If there is no active context, the ROOT_CONTEXT will be returned, which
  // has no key-value pairs.
  // 
  // context.with requires at least 2 arguments: a context and a function to be called.
  // If a third argument is provided, it will be bound to `this` `this` inside the function.
  // Any additional parameters are used as arguments when calling the function.
  // 
  //   Return value is the value returned from getUser
  //    |                         Context to be used as the "active" context
  //    |                         |    Function to be called
  //    |                         |    |        Object assigned to this during function execution
  //    |                         |    |        |          Passed as the first argument to getUser
  //    |                         |    |        |          |
  //    V                         V    V        V          V
  const user = await context.with(ctx, getUser, undefined, userId);
```

Reviewer 2: > I think this calling convention will be the more common and recommended approach most of the time.

I doesn't expect end users to understand this but thats just a personal opinion. 

> What if I annotate the call to context.with like this

I'm not strictly against this way so i'll go and approve the PR, we'll see later on if thats clear enough. Since you already wrote the detailed explaination i think you can add it, its better than nothing

Author: There will be a whole md file dedicated to context as well which will contain multiple different calling conventions and the relative benefits of each.

  // attributes may also be added after the span is started
  span.setAttribute("http.status", 200);

  response.send(user.toJson());
  span.end();

  // Attributes MAY NOT be added after the span ends
  span.setAttribute("my.attribute", false); // this does nothing
}

async function getUser(userId) {
  // when this span is created, it will automatically use the span from the context as its parent
  const span = tracer.startSpan("SELECT ShopDb.Users", {
    attributes: {
      "db.system": "mysql",
      "db.connection_string": "Server=shopdb.example.com;Database=ShopDb;Uid=billing_user;TableCache=true;UseCompression=True;MinimumPoolSize=10;MaximumPoolSize=50;",
      "db.user": "app_user",

>       "net.peer.name": "shopdb.example.com",

Reviewer 2: could we refer to semantic conventions here to explain that those are "standard" attributes ?

  // attributes may also be added after the span is started
  span.setAttribute("http.status", 200);

  response.send(user.toJson());
  span.end();

  // Attributes MAY NOT be added after the span ends
  span.setAttribute("my.attribute", false); // this does nothing
}

async function getUser(userId) {
  // when this span is created, it will automatically use the span from the context as its parent
  const span = tracer.startSpan("SELECT ShopDb.Users", {
    attributes: {
      "db.system": "mysql",
      "db.connection_string": "Server=shopdb.example.com;Database=ShopDb;Uid=billing_user;TableCache=true;UseCompression=True;MinimumPoolSize=10;MaximumPoolSize=50;",
      "db.user": "app_user",
      "net.peer.name": "shopdb.example.com",
      "net.peer.ip": "192.0.2.12",
      "net.peer.port": 3306,
      "net.transport": "IP.TCP",
      "db.name": "ShopDb",
      "db.statement": `Select * from Users WHERE user_id = ${userId}`,
      "db.operation": "SELECT",
      "db.sql.table": "Users",
    },
    kind: SpanKind.CLIENT,
  });
  const user = await db.select("Users", { id: userId });

  span.end();
  return user;
}

server.on("GET", "/user/:id", onGet);
```

### Span Relationships

> One of the most important aspects of spans is their relationships to each other. For instance, if one span describes an incoming request which makes a database call, it may be useful to trace the database call as a separate span which is a child of the original request span. In order to do this, when we create a span we can tell OpenTelemetry which span to use as its parent using a mechanism called _Context_.

Reviewer: ```suggestion
One of the most important aspects of spans is their relationships to each other. For instance, if one span describes an incoming request which makes a database call, it's recommended to trace the database call as a separate span which is a child of the original request span. In order to do this, when we create a span we can tell OpenTelemetry which span to use as its parent using a mechanism called _Context_.
```
